### ISSUE TITLE
Import to Extra instead of Note

### ISSUE THREAD

**Issue by dstark** [retorquere/zotero-better-bibtex#2191]
See support log ID ZYWCXKIG-refs-euc. This probably isn't so much an issue as it is my just not seeing how to handle what I'm trying to do with BBT.

You can access the Zotero item in the support log via the API [here](https://api.zotero.org/users/2212/items/SKXRTSHH?format=bibtex&key=UKV0yq9laZRS6aSpTskUdD1W). The default import though is to attach `annote: BDAG` and other contents of the Extra field as a note on the Zotero record.

What I'm trying to see if I can do with BBT is to store `annote: BDAG` in Zotero in such a way that, if someone else has BBT installed also, adding the citation from the bibtex API link will mean that `annote: BDAG` automatically goes back into (their) Zotero's Extra field without having to be copied and pasted there by the user.

Is this possible? And if so, could you point me in the right direction for how to achieve this result?

Thanks so much!

**Comment by dstark** [comment]
I should perhaps add that "importExtra" is set to "True."

**Comment by retorquere** [comment]
With this build you can set `importNoteToExtra` to value `note`, and the contents of the `note` field will be imported to extra.

There is a remaining issue, that the Zotero bibtex exporter exports your extra field to

```
note = {annote: BDAG
bibtex.note=annote: BDAG},
```

note the single newline. In LaTeX this is equivalent to a space, so that's how it's imported, and you'll end up with a single line in your `extra` field after import.

**Comment by dstark** [comment]
Thanks so much! It looks like this works quite well with the exception one particular version of the multi-line case you mentioned.

Sometimes, it's necessary to add multiple fields in Extra to produce the proper output with Zotero. Here's [one such case](https://api.zotero.org/users/2212/items/HPU37YLY?format=bibtex&key=NeFPXI1BdDlr0bHXmfcIML8S). In several styles, `annote` is being used to store abbreviations that sometimes need to be italicized, sometimes not. So, the CSL is built around roman output and having the `<i> … </i>` tags in Extra with the rest of the content of the `annote` variable when the output needs to be italicized.

I see that [rich text is an exclusion for the importNoteToExtra function you've added](https://retorque.re/zotero-better-bibtex/installation/preferences/hidden-preferences/#importnotetoextra). And that makes a lot of sense given some of what else I've seen about how others are using the Extra field and BBT together. But, correspondingly, when I import [this example reference](https://api.zotero.org/users/2212/items/HPU37YLY?format=bibtex&key=NeFPXI1BdDlr0bHXmfcIML8S) back into Zotero with the new BBT build installed and enabled, I get "Annote: _ISBE_ issued: 1979/1988" appended in a note rather than having it placed back into Extra.

Is/could there be a way to force a bypass of this exclusion in particular cases where rich text tagging needs to be included in Extra?

**Comment by retorquere** [comment]
Why not use RIS or RDF as the share format rather than bibtex?

**Comment by retorquere** [comment]
That's not great. 

Unfortunately, it would be a pretty awkward change to make to my bibtex parser. I can make it so that the extra field roundtrips through BBT, but the stock exporter which exports a single newline won't roundrip that way. And I don't see any translator that roundtrips the `extra` field.

Arguably, this share functionality of Zotero should offer a format that reliably copies items between libraries.

**Comment by dstark** [comment]
> Arguably, this share functionality of Zotero should offer a format that reliably copies items between libraries.

I totally agree. Shared libraries might fit that bill in certain scenarios, but they also have some problematic downsides too.

> I can make it so that the extra field roundtrips through BBT, but the stock exporter which exports a single newline won't roundrip that way.

Does this mean #2 above (round tripping Extra even when rich text tags are applied to variables stored there) is feasible but #1 is the one that might require more awkward changes? 

If not, I totally get that BBT can't be expected to do everything. But if so, having BBT handle cases like #2 above would be a definite boon. And then, users importing the citations should hopefully pretty easily add in the line break on the usually rarer occasion that that's actually needed.

Thanks, again, so much for the thought and time you're giving to this use case!

**Comment by retorquere** [comment]
> Does this mean no. 2 above (round tripping Extra even when rich text tags are applied to variables stored there) is feasible but no. 1 is the one that might require more awkward changes?

There really ought not to be rich text in the extra field. The problem is more that if at import I'm offered a note-like field, I cannot know whether that originated in a plaintext-source (like the `extra` field) or a richtext source (like a note). If there is richtext in the note being offered for import, I have to assume it had a note-like source.

The awkward change is that even if you indicate "whenever you see a note field, assume it is plaintext meant for the `extra` field", I would have to add a special mode to my bibtex parser to say "if you are parsing this field, newlines behave non-latexy". My parser is largely context-free, and adding context is complicated, and this is already a pretty complicated parser. The parser is its own project; BBT sees the content after this parser has done its job, and newlines (aot) will have been transformed already.

I will give it a stab, but it feels really weird to add this to the bibtex parser. It may be possible to use a variant of the verbatim mode.

> If not, I totally get that BBT can't be expected to do everything. But if so, having BBT handle cases like no. 2 above would be a definite boon. And then, users importing the citations should hopefully pretty easily add in the line break on the usually rarer occasion that that's actually needed.

I don't mind doing work for a feature that's rarely used, but this change would add an out-of-spec mode to the bibtex parser solely because bibtex is bein (ab)used as a data exchange format. BBT's aim is to generate functionally appropriate output, and that largely means "if you render a bibliography with Zotero, and from BBT-produced biblatex from the same items, it is a design goal that they look the same".

It is another soft goal that data roundtrips where possible, so that would fit this bill, but for this I've changed how BBT exports the extra field, so that it can be parsed correctly by a compliant parser. The output that the stock exporter yields is understandable, but the fact remains that

```
field = {As you set out for Ithaka
hope the voyage is a long one,
full of adventure, full of discovery.
}
```

and

```
field = {As you set out for Ithaka hope the voyage is a long one, full of adventure, full of discovery.}
```

mean exactly the same in latex.

**Comment by dstark** [comment]
Some option for "whatever's in `extra`, treat as plain text" sounds workable to get tagged text into that Zotero field. Though, if [your idea for an RDF-based mechanism](https://forums.zotero.org/discussion/comment/411104) works out, it sounds like that's a still better way to go given that (1) it doesn't require shoehorning BibTeX and (2) doesn't leave the lingering issue of multiple lines needing to be re-broken out in `extra` again after import.
